luals-addon/factorio/factorio-plugin/util.lua

```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
+
+  local char_escaped = false
+
+  ---check if the next character(s) are equal to the given string
+  ---@param query string
+  ---@return boolean
+  local function peek(query)
+    if cursor + #query > #source then return false end
+    return source:sub(cursor, cursor + #query - 1) == query
+  end
+
+  local function take(query)
+    if not peek(query) then return false end
+    cursor = cursor + #query
+    return true
+  end
+
+  ---Parse a opening long bracket, like `[[` or `[=[`
+  ---Assumes the first bracket has already been consumed.
+  ---@return boolean, integer | nil
+  local function parse_longbracket_open()
+    -- Consume all the '='s
+    local count = 0
+    while take("=") do
+      count = count + 1
+    end
+    if take("[") then
+      return true, count
+    else
+      return false, nil
+    end
+  end
+
+  local function append_range()
+    ranges[#ranges + 1] = {
+      from = start,
+      to = _end,
+      length = _end - start + 1,
+      content = source:sub(start, _end),
+    }
+  end
+
+  local start_clock = os.clock()
+
+  while cursor <= #source do
+    -- read this as a switch statement.
+    local origin = cursor
+
+    ---@type {[LexerState]: fun()}
+    local modes = {
+      code = function()
+        -- rapid advance to the next interesting character
+        cursor = string.match(source, "()[-[\"']", cursor) or #source + 1
+        local anchor = cursor
+        if take("--") then
+          local anchor2 = cursor -- we're still a comment if the long bracket is invalid
+          if take("[") then
+            local is_long, count = parse_longbracket_open()
+            if is_long then
+              if not count then return end
+              state = "long_comment"
+              start = anchor
+              delimit = "]" .. ("="):rep(count) .. "]"
+              return
+            else
+              cursor = anchor2
+            end
+          end
+          state = "short_comment"
+          start = anchor
+        elseif take("[") then
+          local is_long, count = parse_longbracket_open()
+          if is_long then
+            if not count then return end
+            state = "long_string"
+            char_escaped = false
+            start = anchor
+            delimit = "]" .. ("="):rep(count) .. "]"
+          end
+        elseif take('"') then
+          state = "short_string"
+          char_escaped = false
+          start = anchor
+          delimit = '"'
+          patterned_delimit = "()[\\"..delimit.."]"
+        elseif take("'") then
+          state = "short_string"
+          char_escaped = false
+          start = anchor
+          delimit = "'"
+          patterned_delimit = "()[\\"..delimit.."]"
+        else
+          cursor = cursor + 1
+        end
+      end,
+      short_string = function()
+        -- we still need to handle escapes correctly
+        if not char_escaped then
+          cursor = string.match(source, patterned_delimit, cursor) or #source + 1
```

Comment:
For this and the next 3 instances of `string.match`, `string.find` wouldn't require the position capture `()`, but is it faster?

Reply:
I've tested it a bit, there's no measurable difference. So :shrug: other parts are more expensive, basically.